Add Timer Display Window

Caladius opened the pull request:
Using ProxySaw's Notification System as a Style Guide and overall concept, this adds a new Timer Display Window dedicated to On-Screen Timers.

- In-Game Timer, this displays the RTA Style Timer from Gameplay Stats.
- Time of Day, this displays the in-game clock instead of needing to look at the sky or pimpslap a Gossip Stone.

Each Timer has the option to hide the Label and only show the Timer. You can increase the Font Scale for bigger text and hide the Window Background.

More options can be added easily, once the ability to remove the black background from the Digit textures used to draw the current In-Game Timer overlay I will add in those textures to match up. (Once this is done I plan to rip out the source code changes for our current in-game timer in z_parameter.c).

![image](https://github.com/user-attachments/assets/0b2723c1-9d71-41f8-916a-9c3f19491c5f)


<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2257336497.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2257338148.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2257341114.zip)
<!--- section:artifacts:end -->

Caladius committed:
Add new timer window and display options

Caladius committed:
Update to latest dev

Caladius committed:
Fix overflow spacing

Caladius committed:
Icons for Timers

Caladius committed:
Swap Labels for Icons.

Caladius committed:
Additional Timers

Caladius committed:
Navi Timer Icon

Caladius commented:
![image](https://github.com/user-attachments/assets/d62a7ab8-c19e-4839-9ff2-e4e81f527dbb)

Added new Timers and Icon (WIP)

Malkierian commented:
Let us know when you're done building this out and it's ready for review.

Pepper0ni commented:
maybe put it in draft until it's ready to review, so i know when to stop skipping it when skimming the PRs

Caladius committed:
Clean up unused Defines.

Caladius committed:
Merge remote-tracking branch 'origin/develop' into develop-add-timers

Caladius committed:
Code clean up and double check. Ready for Review

Caladius commented:
@Malkierian this is ready for review

inspectredc commented on soh/src/code/z_parameter.c:
```diff
@@ -6056,18 +6056,22 @@ void Interface_Draw(PlayState* play) {
-                        svar5 = CVarGetInteger(CVAR_COSMETIC("HUD.Timers.PosX"), 0)+204+X_Margins_Timer;
-                    } else if (CVarGetInteger(CVAR_COSMETIC("HUD.Timers.PosType"), 0) == 4) {//Hidden
-                        svar5 = -9999;
+                if (CVarGetInteger(CVAR_COSMETIC("HUD.Timers.PosType"), 0) == 4 || 
```
Something mentioning this as the 'hidden' postype i think would help until we have the enums for these

inspectredc commented on soh/src/code/z_parameter.c:
```diff
@@ -6056,18 +6056,22 @@ void Interface_Draw(PlayState* play) {
-                        svar5 = -9999;
+                if (CVarGetInteger(CVAR_COSMETIC("HUD.Timers.PosType"), 0) == 4 || 
+                    (CVarGetInteger(CVAR_WINDOW("TimeDisplayEnabled"), 0) && 
+                        CVarGetInteger(CVAR_ENHANCEMENT("TimeDisplay.Timers.HotWater"), 0) == 1)) {
```
```suggestion
                        CVarGetInteger(CVAR_ENHANCEMENT("TimeDisplay.Timers.HotWater"), 0))) {
```

inspectredc replied:
If this isn't a bool as it looks, then could we have an enum for the states this could be in

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+    Ship::Context::GetInstance()->GetWindow()->GetGui()->LoadGuiTexture("GAMEPLAY_TIMER", gClockIconTex, ImVec4(1, 1, 1, 1));
+    Ship::Context::GetInstance()->GetWindow()->GetGui()->LoadGuiTexture("DAY_TIME_TIMER", gSunIcoTex, ImVec4(1, 1, 1, 1));
+    Ship::Context::GetInstance()->GetWindow()->GetGui()->LoadGuiTexture("NIGHT_TIME_TIMER", gMoonIcoTex, ImVec4(1, 1, 1, 1));
+    Ship::Context::GetInstance()->GetWindow()->GetGui()->LoadGuiTexture("NAVI_TIMER", gNaviIcoTex, ImVec4(1, 1, 1, 1));
```
Could these last 3 be named gXXXXXIconTex instead?

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+		ImGui::BeginTable("Timer List", 2, ImGuiTableFlags_NoClip);
+		for (auto& timers : activeTimers) {
+			ImGui::PushID(timers.timeID);
+            TimeDisplayGetTimer(timers.timeID);
+            ImGui::TableNextColumn();
+            ImGui::Image(textureDisplay, ImVec2(16.0f * fontScale, 16.0f * fontScale));
+            ImGui::TableNextColumn();
+            
+            if (timeDisplayTime != "-:--") {
+                char* textToDecode = new char[timeDisplayTime.size() + 1];
+                textToDecode = std::strcpy(textToDecode, timeDisplayTime.c_str());
+                size_t textLength = timeDisplayTime.length();
+                uint16_t textureIndex = 0;
+            
+                for (size_t i = 0; i < textLength; i++) {
+                    ImVec2 originalCursorPos = ImGui::GetCursorPos();
+                    if (textToDecode[i] == ':' || textToDecode[i] == '.') {
+                        textureIndex = 10;
+                    } else {
+                        textureIndex = textToDecode[i] - 48;
+                    }
+                    if (textToDecode[i] == '.') {
+                        ImGui::SetCursorPosY(ImGui::GetCursorPosY() + (8.0f * fontScale));
+                        ImGui::Image(Ship::Context::GetInstance()->GetWindow()->GetGui()->GetTextureByName(digitList[textureIndex].first),
+                            ImVec2(8.0f * fontScale, 8.0f * fontScale), ImVec2(0, 0.5f), ImVec2(1, 1), textColor, ImVec4(0, 0, 0, 0));
+                    } else {
+                        ImGui::Image(Ship::Context::GetInstance()->GetWindow()->GetGui()->GetTextureByName(digitList[textureIndex].first),
+                            ImVec2(8.0f * fontScale, 16.0f * fontScale), ImVec2(0, 0), ImVec2(1, 1), textColor, ImVec4(0, 0, 0, 0));
+                    }
+                    ImGui::SameLine(0, 0);
+                }
+            }
```
Some weird indentation formatting going on here

Caladius replied:
Not sure where you are referring to here. Looks indented correctly in my editor.

inspectredc replied:
sorry the capture doesnt really show it but compared to the surrounding code its not got enough indentation as a whole block

Malkierian replied:
![image](https://github.com/user-attachments/assets/9bae34db-92f1-4358-a27e-0223768fcb4b)
Perhaps this will clarify what inspect was talking about.

Malkierian replied:
I think this is happening because some of your stuff is using tabs where it's using spaces for whitespace elsewhere. I think the spaces are our standard, so you should do a find and replace for tabs to 4 spaces.

Caladius replied:
makes more sense, I just did an auto-format which should hopefully fix that on my next commit.

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+                    if (textToDecode[i] == ':' || textToDecode[i] == '.') {
+                        textureIndex = 10;
+                    } else {
+                        textureIndex = textToDecode[i] - 48;
```
```suggestion
                        textureIndex = textToDecode[i] - '0';
```

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+            timeDisplayTime = formatTimeDisplay(GAMEPLAYSTAT_TOTAL_TIME).c_str();
+            break;
+        case DISPLAY_TIME_OF_DAY:
+            if (gSaveContext.dayTime >= 17759 && !(gSaveContext.dayTime >= 49155)) {
```
Could the CLOCK_TIME macro be used for these (I know in the codebase we haven't adopted it from decomp yet, but we still have the macro at least)

Caladius replied:
It looks like, with CLOCK_TIME, you still need to feed it an **hr** and a **min** which needs to be determined using the same process I'm using for convertDayTime(), not sure this really helps at all or saves any time/energy?

inspectredc replied:
I meant instead of the 17759 and 49155 numbers, so that its clear what times these mean

Malkierian replied:
I agree, clarification of the time values those represent with the macro would be useful. Also maybe why those values were used, not sure if I'm following the use case.

Caladius replied:
Would something like this work?
![image](https://github.com/user-attachments/assets/78838b67-90f6-493a-989c-bc1728db4fca)

for clarification:
0 - 600 (Navi will spawn at 600)
601 - 3000 (Navi is active)
3001 - 25800 (Navi is gone, the timer resets to 0 when it hits 25800 where Navi can then spawn once more at 600)

Malkierian replied:
That wouldn't be everything you'd add to the enum, right? I think I'd be OK with that for the actual values passed, but I'd still like a human-readable time of day *somewhere* for what those values represent.

Malkierian replied:
Especially since the 17559 and 49155 seem to be actual time values rather than time ranges or durations, like those NAVI ones are.

Caladius replied:
Oops got confused with Navi Timer instead of Time of Day. Yea I can enum the Start of Day and Start of Night values.

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+#define COLOR_LIGHT_GREEN ImVec4(0.52f, 1.0f, 0.23f, 1.0f)
+#define COLOR_GREY ImVec4(0.78f, 0.78f, 0.78f, 1.0f)
+
+std::vector<std::pair<std::string, const char*>> digitList = {
```
```suggestion
const static std::vector<std::pair<std::string, const char*>> digitList = {
```

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+    { DISPLAY_NAVI_TIMER,           "Display Navi Timer",           CVAR_ENHANCEMENT("TimeDisplay.Timers.NaviTimer") }
+};
+
+std::vector<TimeObject> activeTimers;
```
```suggestion
static std::vector<TimeObject> activeTimers;
```

inspectredc commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+    { "COLON_TEXTURE",   gCounterColonTex },
+};
+
+std::vector<TimeObject> timeDisplayList = {
```
```suggestion
const std::vector<TimeObject> timeDisplayList = {
```

inspectredc commented on soh/src/code/z_parameter.c:
```diff
@@ -6056,18 +6056,22 @@ void Interface_Draw(PlayState* play) {
+                    (CVarGetInteger(CVAR_WINDOW("TimeDisplayEnabled"), 0) && 
+                        CVarGetInteger(CVAR_ENHANCEMENT("TimeDisplay.Timers.HotWater"), 0) == 1)) {
+                    svar5 = -9999;
+                } else {
+                    if (CVarGetInteger(CVAR_COSMETIC("HUD.Timers.PosType"), 0) != 0) {
```
I think this reads better if you combine this into an else if

inspectredc reviewed (commented):
Looks good! Just some tidying up comments from me really. Since most these functions are localised to just this file too I'd say either slap static on them or put them in some TimeDisplay:: namespace

Caladius committed:
Update from Suggestions

Caladius committed:
Update magic numbers, correct indentations hopefully

Caladius commented:
Should all be good now?

Malkierian commented on soh/soh/Enhancements/TimeDisplay/TimeDisplay.cpp:
```diff
@@ -0,0 +1,255 @@
+            timeDisplayTime = formatTimeDisplay(GAMEPLAYSTAT_TOTAL_TIME).c_str();
+            break;
+        case DISPLAY_TIME_OF_DAY:
+            if (gSaveContext.dayTime >= DAY_BEGINS && !(gSaveContext.dayTime >= NIGHT_BEGINS)) {
```
One more thing, I think it would make more sense to have this be `gSaveContext.dayTime < NIGHT_BEGINS` rather than this !>=.

Caladius replied:
doneskis

Caladius committed:
One moooooore thing

Pepper0ni reviewed (changes requested):
Issues I found in testing:
- The UI scale lacks +/- buttons
- When Navi decides not to talk to you, the timer is until Navi's variable resets, not when Navi talks to you again, which can be seen if you save edit the timer to 25600.
- the Navi timer displays a very large number when manually save edited above 26000, should probably just display 30 seconds as it will immediately loop around on resume.
- Trade quest timers do not seem to be shown in the new UI, but the existing timer is hidden
- The same goes for the Ganons castle escape timer
- and Running man "Minigame"
- Toggling the various timers causes them to reorder in the window until ship is restarted.

While I don't know how many of these you'd be able to fix, I would at least want unsupported timers to still display normally while the UI is enabled, and will request changes until at least that is fixed.

Caladius committed:
Undo z_param change

Caladius committed:
Updates

Caladius commented:
> Issues I found in testing:
> 
>     * The UI scale lacks +/- buttons
> 
>     * When Navi decides not to talk to you, the timer is until Navi's variable resets, not when Navi talks to you again, which can be seen if you save edit the timer to 25600.
> 
>     * the Navi timer displays a very large number when manually save edited above 26000, should probably just display 30 seconds as it will immediately loop around on resume.
> 
>     * Trade quest timers do not seem to be shown in the new UI, but the existing timer is hidden
> 
>     * The same goes for the Ganons castle escape timer
> 
>     * and Running man "Minigame"
> 
>     * Toggling the various timers causes them to reorder in the window until ship is restarted.
> 
> 
> While I don't know how many of these you'd be able to fix, I would at least want unsupported timers to still display normally while the UI is enabled, and will request changes until at least that is fixed.

1. I'll re-add the +/- buttons but it didnt really make sense to me as it moves in very small increments and you can CTRL+Click to set the value.
2. I'm not entirely concerned about Save Editor changes as those are developer tools. Navi's Enums I added are set to their appropriate "states" and the timer resets to 0 once it hits 25800.
3. I removed the z_parameter changes that auto hide the timer, it will be up to the player to hide it in the Cosmetics Editor.
4. The Timer Order has been corrected here, it will ensure it remains in that order even when toggled.

Pepper0ni commented:
> I'll re-add the +/- buttons but it didnt really make sense to me as it moves in very small increments and you can CTRL+Click to set the value.

Sounds like the issue I hit in the notifications UI that I had to workaround by quadrupling the max scale. This is an issue that needs fixing at some point as it keeps coming up.
